### PR TITLE
Qemu features3

### PR DESCRIPTION
**Pull request by WorksButNotTested**
Also add support for
- [x] Asan
- [x] Cmplog
- [x] Tui
- [x] Verbose (client output)
- [x] Iterations (persistent mode limit)
- [x] include/exclude ranges
- [x] Fix poor performance
- [x] Fix repeated rebuild issues
- [x] Validate core selection (e.g. asan cores must be within cores)

### CODE REVIEW

**Comment by tokatoka** [comment]
can I merge this?

File: fuzzers/qemu_launcher/src/harness.rs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            buf = &buf[0..MAX_INPUT_SIZE];
+            len = MAX_INPUT_SIZE;
+        }
+        let len = len as GuestReg;
+
+        unsafe { self.emu.write_mem(self.input_addr, buf) };
+
+        self.emu
+            .write_reg(Regs::Pc, self.pc)
+            .map_err(|e| Error::unknown(format!("Failed to write PC: {e:}")))?;
```
Comment (domenukk) [inline fuzzers/qemu_launcher/src/harness.rs:65]: I love your detailed error handling (and please do leave it that way), but for the future, probably a `?` could be good enough here? Not sure how verbose the emu error messages are, but maybe we should make hem more verbose if they are not sufficient

File: libafl_qemu/src/elf.rs [hunk 2]
```diff
@@ -67,6 +67,32 @@ impl<'a> EasyElf<'a> {
         None
     }
 
+    #[must_use]
+    pub fn get_section(&self, name: &str, load_addr: GuestAddr) -> Option<Range<GuestAddr>> {
+        for section in &self.elf.section_headers {
+            if let Some(section_name) = self.elf.shdr_strtab.get_at(section.sh_name) {
+                println!(
```
Comment (domenukk) [inline libafl_qemu/src/elf.rs:74]: This probably also wants to use `logger::something`

**Review by domenukk — APPROVED** [review]

File: fuzzers/qemu_launcher/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            } else {
+                Emulator::new(&args, &env)?
+            }
+        };
+
+        let start_pc = Self::start_pc(&emu)?;
+        println!("start_pc @ {start_pc:#x}");
+
+        emu.set_breakpoint(start_pc);
+        unsafe { emu.run() };
```
[the remark this answers is not in the document]
Comment (andreafioraldi) [reply fuzzers/qemu_launcher/src/client.rs:117]: emu.entry_break()

File: fuzzers/qemu_launcher/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+        } else if is_asan {
+            let helpers = tuple_list!(edge_coverage_helper, QemuAsanHelper::default(),);
+            Instance::builder()
+                .options(self.options)
+                .emu(&emu)
+                .mgr(mgr)
+                .core_id(core_id)
+                .build()
+                .run(helpers, state)
+        } else if is_cmplog {
```
[the remark this answers is not in the document]
Comment (andreafioraldi) [reply fuzzers/qemu_launcher/src/client.rs:153]: yes

File: fuzzers/qemu_launcher/Cargo.toml [hunk 5]
```diff
@@ -29,4 +29,15 @@ clap = { version = "4.3.0", features = ["derive", "string"]}
 libafl = { path = "../../libafl/" }
 libafl_bolts = { path = "../../libafl_bolts/" }
 libafl_qemu = { path = "../../libafl_qemu/", features = ["usermode"] }
-rangemap = { version = "1.3" }
+nix = { version = "0.26" }
+rangemap = { version = "1.0.3" }
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/Cargo.toml]: Fixed

File: fuzzers/qemu_launcher/Cargo.toml [hunk 6]
```diff
@@ -29,4 +29,15 @@ clap = { version = "4.3.0", features = ["derive", "string"]}
...
+readonly = { version = "0.2.10" }
+typed-builder = {version = "0.15.1"}
+
+
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/Cargo.toml]: Fixed

File: fuzzers/qemu_launcher/src/client.rs [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+
+        let start_pc = elf
+            .resolve_symbol("LLVMFuzzerTestOneInput", emu.load_addr())
+            .ok_or_else(|| Error::empty_optional("Symbol LLVMFuzzerTestOneInput not found"))?;
+        Ok(start_pc)
+    }
+
+    fn coverage_filter(&self, emu: &Emulator) -> Result<QemuInstrumentationFilter, Error> {
+        if let Some(includes) = &self.options.include {
+            #[allow(clippy::useless_conversion)]
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/client.rs]: It is required for 32-bit target architectures. Updated and commented.

File: fuzzers/qemu_launcher/src/client.rs [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+            let rules = includes
+                .iter()
+                .map(|x| Range {
+                    start: x.start.into(),
+                    end: x.end.into(),
+                })
+                .collect::<Vec<Range<GuestAddr>>>();
+            Ok(QemuInstrumentationFilter::AllowList(rules))
+        } else if let Some(excludes) = &self.options.exclude {
+            #[allow(clippy::useless_conversion)]
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/client.rs]: It is required for 32-bit target architectures. Updated and commented.

File: fuzzers/qemu_launcher/src/client.rs [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    pub fn run(
+        &self,
+        state: Option<ClientState>,
+        mgr: LlmpRestartingEventManager<ClientState, StdShMemProvider>,
+        core_id: CoreId,
+    ) -> Result<(), Error> {
+        let mut args = self.args()?;
+        println!("ARGS: {:#?}", args);
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/client.rs]: Fixed

File: fuzzers/qemu_launcher/src/client.rs [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+
+        let mut env = self.env()?;
+        println!("ENV: {:#?}", env);
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/client.rs]: Fixed

File: fuzzers/qemu_launcher/src/client.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/client.rs]: Fixed

File: fuzzers/qemu_launcher/src/client.rs [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+
+        let edge_coverage_helper = QemuEdgeCoverageHelper::new(self.coverage_filter(&emu)?);
+
+        if is_asan && is_cmplog {
+            let helpers = tuple_list!(
+                edge_coverage_helper,
+                QemuCmpLogHelper::default(),
+                QemuAsanHelper::default(),
+            );
+            Instance::builder()
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/client.rs]: Good idea. That's much nicer! Fixed.

File: fuzzers/qemu_launcher/src/fuzzer.rs [hunk 12]
```diff
@@ -1,315 +1,115 @@
...
-            emu.write_reg(Regs::Sp, stack_ptr)?;
-            emu.write_return_address(ret_addr)?;
-            emu.write_function_argument(CallingConvention::Cdecl, 0, input_addr)?;
-            emu.write_function_argument(CallingConvention::Cdecl, 1, len)?;
-            emu.run();
-            Ok(())
+    pub fn fuzz(&self) -> Result<(), Error> {
+        if self.options.tui {
+            let ui = TuiUI::with_version(
+                String::from("QEMU Launcher For Libpng"),
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/fuzzer.rs]: Fixed

File: fuzzers/qemu_launcher/src/fuzzer.rs [hunk 13]
```diff
@@ -1,315 +1,115 @@
...
+                OpenOptions::new()
+                    .append(true)
+                    .create(true)
+                    .open(l)
+                    .ok()
+                    .map(RefCell::new)
+            });
+
+            #[cfg(unix)]
+            let stdout_cpy = RefCell::new(unsafe {
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/fuzzer.rs]: Fixed

File: fuzzers/qemu_launcher/src/fuzzer.rs [hunk 14]
```diff
@@ -1,315 +1,115 @@
...
-        let edges_observer = unsafe {
-            HitcountsMapObserver::new(VariableMapObserver::from_mut_slice(
-                "edges",
-                edges_map_mut_slice(),
-                addr_of_mut!(MAX_EDGES_NUM),
-            ))
+        let stdout = if self.options.verbose {
+            None
+        } else {
+            Some("/dev/null")
```
[the remark this answers is not in the document]
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/fuzzer.rs:84]: Added a comment to explain.

File: fuzzers/qemu_launcher/src/harness.rs (hunk 1 above)
Comment (WorksButNotTested) [reply fuzzers/qemu_launcher/src/harness.rs:65]: I think I needed to convert the error type maybe?

File: libafl_qemu/src/elf.rs (hunk 2 above)
Comment (WorksButNotTested) [reply libafl_qemu/src/elf.rs]: Fixed

**Pull request merged**